DukeNgn/theia
# Initial contribution of @theia/terminal-external

## What it does

+ Introduces a new extension: `@theia/terminal-external`. The extension adds support for spawning external terminals from Electron applications. In the future, the extension can be extended to support features such as spawning external terminals during debug sessions.

(**Browser applications should not pull this extension**).

#### Command:
+ Supports opening native terminals by calling `Open New External Terminal` from the command palette.

#### Preferences:
+ The PR contributes the following preferences:
	- `terminal.external.windowsExec`
	- `terminal.external.osxExec`
	- `terminal.external.linuxExec`
to allow users to customize the desired terminal application to run on their current platform. 
+ On startup, the backend determines the default terminal on the host machine, then sets the default field in the preference schema accordingly.
+ Changing the `Exec` preference of other OS than the current one won't have any effect. (i.e: Changing `osxExec` while running on Windows doesn't have any effect).

## How to test:
+ Verify the default `exec` of host OS is displaying the correct terminal.

Calling command `Open New External Terminal` has different behaviors, depending on the current workspace status:

#### No workspaces opened and no current active editor -> Spawn the terminal at the user home directory.

https://user-images.githubusercontent.com/43587865/110700880-76183780-81be-11eb-9a89-a85e73fe18e3.mp4


#### No workspaces opened but has an active editor -> Spawn the terminal at the parent folder of the active editor file.

https://user-images.githubusercontent.com/43587865/110700923-83352680-81be-11eb-9f74-2bedbf28c081.mp4


#### Only one workspace opened -> Spawn the terminal at the root of the current workspace.

https://user-images.githubusercontent.com/43587865/110700788-5bde5980-81be-11eb-99c8-2eacd73f0a72.mp4


#### Multi-root workspace opened -> Displays a quick pick to let users choose which workspace to spawn the terminal.

https://user-images.githubusercontent.com/43587865/110700703-40734e80-81be-11eb-8423-d6e72130b811.mp4


## Review checklist


## Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Duc Nguyen <duc.a.nguyen@ericsson.com>
